les/lespay/client/requestbasket.go

```diff
@@ -0,0 +1,285 @@
+// constant so that each server basket represents long term usage and price statistics.
+// When the transferred part is added to the reference basket the values are scaled so
+// that their sum equals the total value calculated according to the previous reqValues.
+// The ratio of request values coming from the server basket represent the pricing of
+// the specific server and modify the global estimates with a weight proportional to
+// the amount of service provided by the server.
+type serverBasket struct {
+	basket   requestBasket
+	rvFactor float64
+}
+
+type (
+	// requestBasket holds amounts and values for each request type.
+	// These values are exponentially expired (see utils.ExpiredValue). The power of 2
+	// exponent is applicable to all values within.
+	requestBasket struct {
+		items []basketItem
+		exp   uint64
+	}
+	// basketItem holds amount and value for a single request type. Value is the total
+	// relative request value accumulated for served requests while amount is the counter
+	// for each request type.
+	// Note that these values are both scaled up by basketFactor because of the exponential
+	// expiration.
+	basketItem struct {
+		amount, value uint64
+	}
+)
+
+// setExp sets the power of 2 exponent of the structure, scaling base values (the amounts
+// and request values) up or down if necessary.
+func (b *requestBasket) setExp(exp uint64) {
+	if exp > b.exp {
+		shift := exp - b.exp
+		for i, item := range b.items {
+			item.amount >>= shift
+			item.value >>= shift
+			b.items[i] = item
+		}
+		b.exp = exp
+	}
+	if exp < b.exp {
+		shift := b.exp - exp
+		for i, item := range b.items {
+			item.amount <<= shift
+			item.value <<= shift
+			b.items[i] = item
+		}
+		b.exp = exp
+	}
+}
+
+// init initializes a new server basket with the given service vector size (number of
+// different request types)
+func (s *serverBasket) init(size int) {
+	if s.basket.items == nil {
+		s.basket.items = make([]basketItem, size)
+	}
+}
+
+// add adds the give type and amount of requests to the basket. Cost is calculated
+// according to the server's own cost table.
+func (s *serverBasket) add(reqType, reqAmount uint32, reqCost uint64, expFactor utils.ExpirationFactor) {
+	s.basket.setExp(expFactor.Exp)
+	i := &s.basket.items[reqType]
+	i.amount += uint64(float64(uint64(reqAmount)*basketFactor) * expFactor.Factor)
+	i.value += uint64(float64(reqCost) * s.rvFactor * expFactor.Factor)
+}
+
+// updateRvFactor updates the request value factor that scales server costs into the
+// local value dimensions.
+func (s *serverBasket) updateRvFactor(rvFactor float64) {
+	s.rvFactor = rvFactor
+}
+
+// transfer decreases amounts and values in the basket with the given ratio and
+// moves the removed amounts into a new basket which is returned and can be added
+// to the global reference basket.
+func (s *serverBasket) transfer(ratio float64) requestBasket {
+	res := requestBasket{
```

Comment:
No because `transfer` keeps the expired format, the transferred basket also inherits the `exp` field. The final readout functions do apply the expiration.

Reply:
But what's the final readout function?

Reply:
`func (vt *ValueTracker) RequestStats() []RequestStatsItem` (which is accessible through the API) does directly read out the amounts from the basket and it indeed applies expiration. Otherwise the basket is not exposed directly, what we use for calculations are ratios (like `value`/`amount` or `amount`/`sumAmount`) where we can just directly use the base values without caring about the expiration.